--- gtkdoc-scangobj	2004-04-13 22:52:13.000000000 +0530
+++ gtkdoc-scangobj.new	2004-04-13 22:52:02.000000000 +0530
@@ -879,7 +879,7 @@
 static void
 output_widget_args (FILE *fp, GType object_type)
 {
-  gpointer class;
+  gpointer klass;
   const gchar *object_class_name;
   gint arg;
   gchar flags[16], *pos;
@@ -890,21 +890,21 @@
 
   if (G_TYPE_IS_CLASSED (object_type))
     {
-      class = g_type_class_peek (object_type);
-      if (!class)
+      klass = g_type_class_peek (object_type);
+      if (!klass)
 	return;
 
-      properties = g_object_class_list_properties (class, &n_properties);
+      properties = g_object_class_list_properties (klass, &n_properties);
     }
 #if GLIB_MAJOR_VERSION > 2 || (GLIB_MAJOR_VERSION == 2 && GLIB_MINOR_VERSION >= 3)
   else if (G_TYPE_IS_INTERFACE (object_type))
     {
-      class = g_type_default_interface_ref (object_type);
+      klass = g_type_default_interface_ref (object_type);
 
-      if (!class)
+      if (!klass)
 	return;
 
-      properties = g_object_interface_list_properties (class, &n_properties);
+      properties = g_object_interface_list_properties (klass, &n_properties);
     }
 #endif
   else
@@ -958,8 +958,8 @@
     g_free (properties);
 
 #ifdef GTK_IS_CONTAINER_CLASS
-    if (!child_prop && GTK_IS_CONTAINER_CLASS (class)) {
-      properties = gtk_container_class_list_child_properties (class, &n_properties);
+    if (!child_prop && GTK_IS_CONTAINER_CLASS (klass)) {
+      properties = gtk_container_class_list_child_properties (klass, &n_properties);
       child_prop = TRUE;
       continue;
     }
@@ -967,8 +967,8 @@
 
 #ifdef GTK_IS_WIDGET_CLASS
 #if GTK_CHECK_VERSION(2,1,0)
-    if (!style_prop && GTK_IS_WIDGET_CLASS (class)) {
-      properties = gtk_widget_class_list_style_properties (GTK_WIDGET_CLASS (class), &n_properties);
+    if (!style_prop && GTK_IS_WIDGET_CLASS (klass)) {
+      properties = gtk_widget_class_list_style_properties (GTK_WIDGET_CLASS (klass), &n_properties);
       style_prop = TRUE;
       continue;
     }
