Index: ScintillaGTK.cxx
===================================================================
RCS file: /cvs/gnome/anjuta/scintilla/ScintillaGTK.cxx,v
retrieving revision 1.21
retrieving revision 1.22
diff -u -r1.21 -r1.22
--- ScintillaGTK.cxx	1 Feb 2005 16:18:56 -0000	1.21
+++ ScintillaGTK.cxx	11 Mar 2005 11:36:43 -0000	1.22
@@ -585,13 +585,15 @@
 	if (sciThis->ic)
 		gdk_im_begin(sciThis->ic, widget->window);
 #else
-	gchar *str;
+	gchar *str = NULL;
 	gint cursor_pos;
 	gtk_im_context_get_preedit_string(sciThis->im_context, &str, NULL, &cursor_pos);
-	if (strlen(str) > 0){
-		gtk_widget_show(PWidget(sciThis->wPreedit));
-	} else{
-		gtk_widget_hide(PWidget(sciThis->wPreedit));
+	if (PWidget(sciThis->wPreedit) != NULL) { 
+		if ((str != NULL) && (strlen(str) > 0)){
+			gtk_widget_show(PWidget(sciThis->wPreedit));
+		} else{
+			gtk_widget_hide(PWidget(sciThis->wPreedit));
+		}
 	}
 	g_free(str);
 	gtk_im_context_focus_in(sciThis->im_context);
@@ -611,8 +613,10 @@
 #if GTK_MAJOR_VERSION < 2
 	gdk_im_end();
 #else
-	gtk_widget_hide(PWidget(sciThis->wPreedit));
-	gtk_im_context_focus_out(sciThis->im_context);
+	if (PWidget(sciThis->wPreedit) != NULL ) {
+		gtk_widget_hide(PWidget(sciThis->wPreedit));
+		gtk_im_context_focus_out(sciThis->im_context);
+	}
 #endif
 #endif
 
