diff -u scintilla/Editor.cxx scintilla-new/Editor.cxx
--- scintilla/PlatGTK.cxx	2003-09-18 12:05:56.000000000 +0530
+++ scintilla-new/PlatGTK.cxx	2003-09-17 11:31:58.000000000 +0530
@@ -494,16 +494,16 @@
 	charset[0] = '\0';
 
 #ifdef USE_PANGO
-	if (fontName[0] == '!') {
-		PangoFontDescription *pfd = pango_font_description_new();
-		if (pfd) {
-			pango_font_description_set_family(pfd, fontName+1);
-			pango_font_description_set_size(pfd, size * PANGO_SCALE);
-			pango_font_description_set_weight(pfd, bold ? PANGO_WEIGHT_BOLD : PANGO_WEIGHT_NORMAL);
-			pango_font_description_set_style(pfd, italic ? PANGO_STYLE_ITALIC : PANGO_STYLE_NORMAL);
-			return new FontHandle(pfd, characterSet);
-		}
+	//if (fontName[0] == '!') {
+	PangoFontDescription *pfd = pango_font_description_new();
+	if (pfd) {
+		pango_font_description_set_family(pfd, fontName);
+		pango_font_description_set_size(pfd, size * PANGO_SCALE);
+		pango_font_description_set_weight(pfd, bold ? PANGO_WEIGHT_BOLD : PANGO_WEIGHT_NORMAL);
+		pango_font_description_set_style(pfd, italic ? PANGO_STYLE_ITALIC : PANGO_STYLE_NORMAL);
+		return new FontHandle(pfd, characterSet);
 	}
+	//}
 #endif
 
 	GdkFont *newid = 0;
