Index: PlatGTK.cxx
===================================================================
RCS file: /cvs/gnome/anjuta/scintilla/PlatGTK.cxx,v
retrieving revision 1.22
diff -u -r1.22 PlatGTK.cxx
--- PlatGTK.cxx	19 Oct 2004 19:21:42 -0000	1.22
+++ PlatGTK.cxx	3 Mar 2005 18:47:23 -0000
@@ -1716,8 +1716,8 @@
 		cursorLast = cursorArrow;
 		break;
 	}
-
-	gdk_window_set_cursor(PWidget(id)->window, gdkCurs);
+	if (PWidget(id)->window)
+	    gdk_window_set_cursor(PWidget(id)->window, gdkCurs);
 	gdk_cursor_destroy(gdkCurs);
 }
 
