--- scintilla/include/Scintilla.h
+++ scintilla/include/Scintilla.h
@@ -24,6 +24,11 @@
 #if _MSC_VER >= 1300
 #include <BaseTsd.h>
 #endif
+#ifdef HAVE_STDINT_H
+#include <stdint.h>
+typedef uintptr_t uptr_t;
+typedef intptr_t sptr_t;
+#else
 #ifdef MAXULONG_PTR
 typedef ULONG_PTR uptr_t;
 typedef LONG_PTR sptr_t;
@@ -31,6 +36,7 @@
 typedef unsigned long uptr_t;
 typedef long sptr_t;
 #endif
+#endif
 
 typedef sptr_t (*SciFnDirect)(sptr_t ptr, unsigned int iMessage, uptr_t wParam, sptr_t lParam);
 
